omnigibson/utils/motion_planning_utils.py

```diff
@@ -159,22 +192,55 @@ def detect_robot_collision(robot, filter_objs=[]):
         if col_obj.category in filter_categories:
             collision_prims.remove(col_prim)
 
-    return len(collision_prims) > 0 or detect_self_collision(robot)
-
-def detect_self_collision(robot):
-    contacts = robot.contact_list()
-    robot_links = [link.prim_path for link in robot.links.values()]
-    disabled_pairs = [set(p) for p in robot.disabled_collision_pairs]
-    for c in contacts:
-        link0 = c.body0.split("/")[-1]
-        link1 = c.body1.split("/")[-1]
-        if {link0, link1} not in disabled_pairs and c.body0 in robot_links and c.body1 in robot_links:
-            return True
-    return False
+    return len(collision_prims) > 0
+    
 
-def detect_hand_collision(robot, joint_pos, control_idx):
-    robot.set_joint_positions(joint_pos, control_idx)
-    return detect_robot_collision(robot)
+# Sets joint positions of robot and detects robot collisions with the environment and itself
+def arm_planning_validity_fn(context, joint_pos):
+    arm_links = context.robot.manipulation_link_names
+    link_poses = context.fk_solver.get_link_poses(joint_pos, arm_links)
+
+    for link in arm_links:
+        pose = link_poses[link]
+        if link in context.robot_meshes_copy.keys():
+            for mesh, relative_pose in zip(context.robot_meshes_copy[link], context.robot_meshes_relative_poses[link]):
+                mesh_pose = T.pose_transform(*pose, *relative_pose)
+                translation = Gf.Vec3d(*np.array(mesh_pose[0], dtype=float))
+                mesh._prim.GetAttribute("xformOp:translate").Set(translation)
+                orientation = np.array(mesh_pose[1], dtype=float)[[3, 0, 1, 2]]
+                mesh._prim.GetAttribute("xformOp:orient").Set(Gf.Quatd(*orientation))
+
+    # Define function for checking overlap
+    valid_hit = False
+    mesh_hit = None
+
+    def overlap_callback(hit):
+        nonlocal valid_hit
+        nonlocal mesh_hit
+        
+        valid_hit = hit.rigid_body not in context.disabled_collision_pairs_dict[mesh_hit]
+
+        return not valid_hit
+
+    def check_overlap():
+        nonlocal valid_hit
+        nonlocal mesh_hit
+        valid_hit = False
+
+        for link in context.robot_meshes_copy:
+            for mesh in context.robot_meshes_copy[link]:
+                if valid_hit:
+                    return valid_hit
+                mesh_id = PhysicsSchemaTools.encodeSdfPath(mesh.prim_path)
+                mesh_hit = mesh.prim_path
+                if mesh._prim.GetTypeName() == "Mesh":
+                    og.sim.psqi.overlap_mesh(*mesh_id, reportFn=overlap_callback)
+                else:
+                    og.sim.psqi.overlap_shape(*mesh_id, reportFn=overlap_callback)
+            
+        return valid_hit
```

Comment:
We can unwrap this function now and directly execute this code rather than calling it. It'll remove some of the annoying logic around the nonlocal stuff. Also, do the `if valid hit` part *after* the overlap call, not before. Also the logic around `mesh_hit` is very confusing too.